### PR TITLE
Add path formation at form::showdocuments to handle custom module

### PR DESCRIPTION
**Pull request by davidNDU**
# New [*Method Form::showdocuments fails with custom module*]
[*Following the various instructions and the code generated from modul builder, it is neccessary to modify _showdocuments_ so it works with a custom module.*]

### CODE REVIEW

File: htdocs/core/class/html.formfile.class.php [hunk 1]
```diff
@@ -628,9 +628,17 @@ public function showdocuments($modulepart, $modulesubdir, $filedir, $urlsource,
 				// For normalized external modules.
 				else {
 				    $file = dol_buildpath('/'.$modulepart.'/core/modules/'.$modulepart.'/modules_'.$submodulepart.'.php', 0);
-					$res = include_once $file;
+					if (file_exists($file))
```
Comment (eldy) [inline htdocs/core/class/html.formfile.class.php:631]: Why do you need this.
With the 
$file = dol_buildpath('/'.$modulepart.'/core/modules/'.$modulepart.'/modules_'.$submodulepart.'.php', 0);
$res = include_once $file;

You can already include the class ModulePDFyourmodulename that can hae a method liste_modeles() to return list of document templates.

**Comment by davidNDU** [comment]
1) File name of PDF document template
As I understood, the file name has to follow the form: 
\MyModuleName\core\modules\MyModuleName\doc\pdf_MyModuleName_ReportName.modules.php
This is according to https://wiki.dolibarr.org/index.php/Create_a_PDF_document_template#Create_your_new_template

2) Class name of PDF template
Again from the wiki entry I understood the class name needs to follow the pattern 'pdf_MyModuleName_ReportName'

With the following settings:
* name of file: pdf_MyModuleName_ReportName.modules.php (specific: pdf_lims_testreport.modules.php)
* class name: Mod**e**lePDFMyModuleName  (specific: ModelePDFlims)
* Constructor class: $this->name = "testreport";
* Entry in llx_document_model: nom=>'testreport', type=>'lims'

There are the following errors:
* commonGenerateDocument (used with actions_addupdatedelete.inc.php), line 4571: $classname = $prefix.'_'.$modele;
=> Class 'pdf_lims_testreport' not found 
* or when used with GENERATE button at $formfile->showdocuments)
=> File C:/wamp64/www/dolibarr/htdocs/ticket/core/modules/lims/doc/pdf_testreport.modules.php does not exist

When I change the name of the class to pdf_lims_testreport the error with GENERATE persists, but the rest works out fine:
Error File C:/wamp64/www/dolibarr/htdocs/ticket/core/modules/lims/doc/pdf_testreport.modules.php does not exist

EDIT:
with the proposed PR in place and the name of the testreport following the convention 'MyModuleName'_'ReportName' (here specific lims_testreport) the GENERATE button works out.

**Comment by eldy** [comment]
I see. The problem is larger than that. Support for document generation for custom modules was clearly not yet complete in v11.
A big work has been done and is ready for v12 (generation of files and templates are now included into generation by modulebuilder). So i keep this PR open for the moment until we are sure that changes in v12 are ok and that this PR is no more required (it appears to be yes according to test i done this week).

**Comment by davidNDU** [comment]
@eldy Thank you for your time and efforts!

**Comment by davidNDU** [comment]
@eldy How do I rectify my module? Do I have to run the module builder again?

**Comment by davidNDU** [comment]
With Dolibarr 13.0.0-alpha 25.9.2020 (d9e7cacc840cbb722511f11c8f026b8f9917b04a):
Module Builder generates code that will result in this error:
grafik

pdf_standard_MyObject.modules.php Line 902
`if ($object->statut == Facture::STATUS_DRAFT && (!empty($conf->global->FACTURE_DRAFT_WATERMARK)))`

pdf_standard_MyObject.modules.php Line 962
`if ($object->statut == Facture::STATUS_DRAFT)`

At both cases `Facture::STATUS_DRAFT` could be changed to `$object::STATUS_DRAFT`

Then after the document is generated, an error comes up because the sql-table misses '**last_main_doc**'
grafik

So the field list should have an entry similar to this
`'last_main_doc' => array('type'=>'varchar(255)', 'label'=>'LastMainDoc', 'enabled'=>1, 'position'=>600, 'notnull'=>0, 'visible'=>0,),`
and 
`public $last_main_doc;`

**Comment by davidNDU** [comment]
#14871 should close this issue

**Pull request closed without merging**